x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java

```diff
@@ -1266,6 +1284,241 @@ private boolean foldable(Expression e) {
 
     }
 
+    public static class ConvertTypes extends OptimizerExpressionRule<ScalarFunction> {
+
+        public ConvertTypes() {
+            super(TransformDirection.DOWN);
+        }
+
+        @Override
+        public Expression rule(ScalarFunction e) {
+            if (e instanceof BinaryComparison predicate && predicate.right().foldable()) {
+                return convertBinaryComparison(predicate);
+            } else if (e instanceof Range range) {
+                return convertRange(range);
+            } else if (e instanceof In in) {
+                return convertIn(in);
+            }
+            return e;
+        }
+
+        private Expression convertIn(In in) {
+            if (in.list().isEmpty()) {
+                return in;
+            }
+            List<Expression> newRight = new ArrayList<>();
+            List<Expression> additionalConditions = new ArrayList<>();
+            DataType type = in.value().dataType();
+            boolean converted = false;
+            for (Expression exp : in.list()) {
+                if (type == exp.dataType() || exp.foldable() == false || DataTypes.areCompatible(type, exp.dataType())) {
+                    newRight.add(exp);
+                } else {
+                    converted = true;
+                    Object foldedValue = exp.fold();
+                    if (DataTypes.isDateTime(type) && DataTypes.isString(exp.dataType())) {
+                        try {
+                            // try date math first
+                            Tuple<Long, Long> range = DateFieldMapper.DateFieldType.toTimestampRange(
+                                foldedValue,
+                                foldedValue,
+                                true,
+                                true,
+                                ZoneOffset.UTC,
+                                DateUtils.DATE_MATH_PARSER,
+                                DateFieldMapper.Resolution.MILLISECONDS,
+                                () -> System.currentTimeMillis()
+                            );
+                            if (range.v1() < range.v2()) {
+                                /*
+                                treat it as a date range
+                                eg. the following
+                                something IN (<daterange>, <somethingElse1>, <somethingElse2>)
+                                becomes
+                                something BETWEEN <min(daterange)> AND <max(daterange)> OR someting IN (<somethingElse1>, <somethingElse2>)
+                                */
+                                additionalConditions.add(
+                                    new Range(
+                                        in.source(),
+                                        in.value(),
+                                        tryCast(exp, range.v1(), type),
+                                        true,
```

Comment:
Should the lower bound be included, if changed (i.e. rounded down)?

Reply:
I'm preserving the semantics of BETWEEN operator here, ie. upper and lower bound are included.
In this case it's just a bit more convoluted, since I have to take into consideration that the date expression can be a date math, so including it all could mean including the range upper or lower bound, if the expression is the first or the second value of the BETWEEN